sycl/source/backend.cpp

```diff
@@ -94,6 +103,96 @@ __SYCL_EXPORT event make_event(pi_native_handle NativeHandle,
+        Plugin.call<PiApiKind::piProgramCompile>(
+            PiProgram, 1, &Dev, nullptr, 0, nullptr, nullptr, nullptr, nullptr);
+      else if (State == bundle_state::executable)
+        Plugin.call<PiApiKind::piProgramBuild>(PiProgram, 1, &Dev, nullptr,
+                                               nullptr, nullptr);
+      break;
+    case (PI_PROGRAM_BINARY_TYPE_COMPILED_OBJECT):
+    case (PI_PROGRAM_BINARY_TYPE_LIBRARY):
+      if (State == bundle_state::input)
+        // TODO SYCL2020 exception
+        throw sycl::runtime_error("Program and kernel_bundle state mismatch",
+                                  PI_INVALID_VALUE);
+      if (State == bundle_state::executable)
+        Plugin.call<PiApiKind::piProgramLink>(ContextImpl->getHandleRef(), 1,
+                                              &Dev, nullptr, 1, &PiProgram,
+                                              nullptr, nullptr, &PiProgram);
+      break;
+    case (PI_PROGRAM_BINARY_TYPE_EXECUTABLE):
+      if (State == bundle_state::input || State == bundle_state::object)
+        // TODO SYCL2020 exception
+        throw sycl::runtime_error("Program and kernel_bundle state mismatch",
+                                  PI_INVALID_VALUE);
+      break;
+    }
+  }
+
+  std::vector<device> Devices;
+  Devices.reserve(ProgramDevices.size());
+  std::transform(
+      ProgramDevices.begin(), ProgramDevices.end(), std::back_inserter(Devices),
+      [&Plugin](const auto &Dev) {
+        auto Platform =
+            detail::platform_impl::getPlatformFromPiDevice(Dev, Plugin);
+        auto DeviceImpl = Platform->getOrMakeDeviceImpl(Dev, Platform);
+        return createSyclObjFromImpl<device>(DeviceImpl);
+      });
+
+  // Unlike SYCL, other backends, like OpenCL or Level Zero, may not support
+  // getting kernel IDs before executable is built. The SYCL Runtime workarounds
+  // this by pre-building the device image and extracting kernel info. We can't
+  // do the same to user images, since they may contain references to undefined
+  // symbols (e.g. when kernel_bundle is supposed to be joined with another).
+  std::vector<kernel_id> KernelIDs{};
+  auto DevImgImpl = std::make_shared<device_image_impl>(
+      nullptr, TargetContext, Devices, State, KernelIDs, PiProgram);
+  device_image_plain DevImg{DevImgImpl};
+
+  return std::make_shared<kernel_bundle_impl>(TargetContext, Devices, DevImg);
+}
+kernel make_kernel(pi_native_handle NativeHandle,
+                   const context &TargetContext) {
+  const auto &Plugin = pi::getPlugin<backend::opencl>();
+  const auto &ContextImpl = getSyclObjImpl(TargetContext);
+  // Create PI kernel first.
+  pi::PiKernel PiKernel = nullptr;
+  Plugin.call<PiApiKind::piextKernelCreateWithNativeHandle>(
+      NativeHandle, ContextImpl->getHandleRef(), &PiKernel);
+  // Construct the SYCL queue from PI queue.
+  return detail::createSyclObjFromImpl<kernel>(
+      std::make_shared<kernel_impl>(PiKernel, ContextImpl));
```

Comment:
I think this is missing the required "retain" for OpenCL backend:
https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_interoperability_with_kernels
That's why I said that "reference counting" makes implementation different for different backends.

Reply:
Done